### PR TITLE
[TEST] ASoC: rt5682: move DAI clock registration to the i2c probe

### PR DESCRIPTION
**Pull request by shumingfan**
[PATCH v2] ASoC: rt5682: move DAI clock registration to the i2c probe

The DAI clock only uses in the i2s mode.
Therefore, moving DAI clock registration to the i2c probe would be right.
We try to avoid the resource release issue with i2s and soundwire mode.

Signed-off-by: Shuming Fan <shumingf@realtek.com>

### CODE REVIEW

**Comment by plbossart** [comment]
Doesn't seem to work @shumingfan, see https://sof-ci.01.org/linuxpr/PR3130/build6262/devicetest/?model=CML_HEL_RT5682&testcase=verify-pcm-list

````
kernel: [    2.681876] rt5682 i2c-10EC5682:00: sysclk/dai not set correctly
kernel: [    2.681982] general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000 [#1] SMP NOPTI
kernel: [    2.681993] CPU: 0 PID: 98 Comm: kworker/u16:2 Not tainted 5.14.0-rc7-pr3130-6262-default #76cfd08f
kernel: [    2.682001] Hardware name: Google Helios/Helios, BIOS  01/21/2020
kernel: [    2.682006] Workqueue: events_unbound async_run_entry_fn
kernel: [    2.682014] RIP: 0010:clk_core_get_parent_by_index+0x4a/0x90
kernel: [    2.682020] Code: 8d 2c 92 48 c1 e5 03 4c 8d 24 28 49 8b 44 24 08 48 85 c0 74 0c 5b 5d 41 5c c3 5b 31 c0 5d 41 5c c3 49 8b 04 24 48 85 c0 74 26 <48> 8b 00 48 85 c0 74 e3 48 3d 00 f0 ff ff 77 05 49 89 44 24 08 48
kernel: [    2.682030] RSP: 0018:ffffa36a00897c08 EFLAGS: 00010286
kernel: [    2.682035] RAX: d7a152bd29b38e00 RBX: ffff90cf1f15a100 RCX: 00000000b815464c
kernel: [    2.682040] RDX: 0000000000000000 RSI: 0000000000000000 RDI: ffff90cf1f15a100
kernel: [    2.682046] RBP: 0000000000000000 R08: 00000000ffffffff R09: 0000000000000000
kernel: [    2.682052] R10: 0000000000000000 R11: ffff90d01431956f R12: ffff90cf142bec40
kernel: [    2.682057] R13: ffff90cf14319f00 R14: ffff90cf142bec68 R15: ffff90cf1f15a100
kernel: [    2.682063] FS:  0000000000000000(0000) GS:ffff90d055600000(0000) knlGS:0000000000000000
kernel: [    2.682070] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
kernel: [    2.682076] CR2: 00007f25280491e8 CR3: 000000010da26003 CR4: 00000000003706f0
kernel: [    2.682082] Call Trace:
kernel: [    2.682087]  __clk_register+0x465/0x7e0
kernel: [    2.682094]  ? clk_hw_unregister+0x10/0x10
kernel: [    2.682100]  clk_hw_register+0x19/0x40
kernel: [    2.682106]  devm_clk_hw_register+0x41/0x80
kernel: [    2.682113]  rt5682_register_dai_clks+0x8e/0x130 [snd_soc_rt5682]
kernel: [    2.682125]  rt5682_i2c_probe+0x484/0x600 [snd_soc_rt5682_i2c]
kernel: [    2.682134]  ? rt5682_irq+0x40/0x40 [snd_soc_rt5682_i2c]
kernel: [    2.682142]  i2c_device_probe+0x314/0x340
kernel: [    2.682149]  really_probe+0x1a6/0x3a0
kernel: [    2.682155]  __driver_probe_device+0xf9/0x170
kernel: [    2.682160]  driver_probe_device+0x19/0x90
kernel: [    2.682165]  __driver_attach_async_helper+0x26/0x70
kernel: [    2.682171]  async_run_entry_fn+0x26/0x110
kernel: [    2.682178]  process_one_work+0x286/0x4f0
kernel: [    2.682184]  worker_thread+0x48/0x3d0
kernel: [    2.682190]  ? process_one_work+0x4f0/0x4f0
kernel: [    2.682195]  kthread+0x141/0x170
kernel: [    2.682201]  ? set_kthread_struct+0x40/0x40
kernel: [    2.682208]  ret_from_fork+0x1f/0x30
````

**Comment by shumingfan** [comment]
SOFCI TEST

**Comment by shumingfan** [comment]
@plbossart @bardliao
I updated the PR which adds more debug messages to show what I observed.
I think the problem might come from the compiler issue.
I try to explain that below.

I can't reproduce the 'general protection fault' issue before.
Since I got the Heilos from Intel, I checked the kernel log and tested with the same kernel configuration.
The different part only is the compiler version and I tested with two different compiler versions.
The PR still works well when building with GCC 5.4.0, however, 
the I built with GCC 9.3.0 which is the same with CI test that the 'general protection fault' sometimes happened.

Details:
The 'general protection fault' issue happened when registered the clock of RT5682_DAI_BCLK_IDX.
The WCLK clock was registered done before the BCLK and WCLK to be parents of BCLK.
In the registering process, it will access the parent of BCLK to set up.
The problem is the code can't get the right pointer address.

Somehow, I add more debug messages, then the problem can't reproduce.
Here, I show three different test logs below. 
I think it needs the compiler expert to help. 
Why the code can't get the proper address, but the old compiler version can?
And the 'general protection fault' message is from the traps.c.
https://github.com/thesofproject/linux/blob/f9860abc87cebc21aee1ac5d1845dc30282a3239/arch/x86/kernel/traps.c#L592

# test with GCC 5.4.0
```
[    3.188966] rt5682_register_dai_clks, i=0, init.name=rt5682-dai-wclk 
[    3.188979] rt5682 i2c-10EC5682:00: sysclk/dai not set correctly
[    3.189072] rt5682_register_dai_clks, dai_clk_hw->core=0x000000009a4090f9 // wclk registerd done and allocated the clk_core.
[    3.189078] rt5682_register_dai_clks, i=1, init.name=rt5682-dai-bclk 
[    3.189081] rt5682_register_dai_clks, RT5682_DAI_WCLK_IDX hw->core=0x000000009a4090f9
[    3.189088] clk_core_get_parent_by_index, before clk_core_fill_parent_index, index=0
[    3.189090] clk_core_fill_parent_index, entry->hw=0x00000000b06e487c 
[    3.189093] clk_core_fill_parent_index, entry->hw->core=0x000000009a4090f9 // got the parent of BCLK, it is the clk_core of WCLK, and the pointer address is right
[    3.189095] clk_core_get_parent_by_index, after clk_core_fill_parent_index
[    3.189197] rt5682_register_dai_clks, dai_clk_hw->core=0x00000000e3629053
[    3.189431] cml_rt1011_rt5682 cml_rt1011_rt5682: sof_rt1011_quirk = f
```
# test with GCC 9.3.0
```
[    2.934973] rt5682_register_dai_clks, i=0, init.name=rt5682-dai-wclk 
[    2.934997] rt5682 i2c-10EC5682:00: sysclk/dai not set correctly
[    2.935167] rt5682_register_dai_clks, dai_clk_hw->core=0x00000000294245c1
[    2.935178] rt5682_register_dai_clks, i=1, init.name=rt5682-dai-bclk 
[    2.935182] rt5682_register_dai_clks, RT5682_DAI_WCLK_IDX hw->core=0x00000000294245c1
[    2.935193] clk_core_get_parent_by_index, before clk_core_fill_parent_index, index=0
[    2.935197] clk_core_fill_parent_index, entry->hw=0x0000000079ebded0 
[    2.935213] general protection fault, probably for non-canonical address 0x62a81aeb8ce3bb00: 0000 [#1] SMP NOPTI // fail to get the clk_core of parent of BCLK
[    2.935237] CPU: 1 PID: 98 Comm: kworker/u16:2 Not tainted 5.14.0-rc7+ #59
[    2.935253] Hardware name: Google Helios/Helios, BIOS  01/21/2020
[    2.935265] Workqueue: events_unbound async_run_entry_fn
[    2.935285] RIP: 0010:clk_core_get_parent_by_index.part.0+0x6a/0xd4
[    2.935305] Code: 48 85 d2 74 37 48 c7 c6 40 b7 8a 98 48 c7 c7 70 67 bd 98 e8 ac e3 fd ff 49 8b 45 00 48 c7 c6 40 b7 8a 98 48 c7 c7 85 67 bd 98 <48> 8b 18 48 89 da e8 8f e3 fd ff 48 85 db 75 27 eb 32 44 89 f6 4c
[    2.935333] RSP: 0018:ffffb26d40897bf0 EFLAGS: 00010246
[    2.935345] RAX: 62a81aeb8ce3bb00 RBX: ffff92d9c399d3d0 RCX: 0000000000000027
[    2.935356] RDX: 0000000000000000 RSI: ffffffff988ab740 RDI: ffffffff98bd6785
[    2.935367] RBP: 0000000000000000 R08: ffff92db159d7770 R09: ffffb26d40897a28
[    2.935379] R10: 0000000000000001 R11: 0000000000000001 R12: ffff92d9c6664600
[    2.935388] R13: ffff92d9c867a400 R14: 0000000000000000 R15: ffff92d9c6664600
[    2.935399] FS:  0000000000000000(0000) GS:ffff92db15800000(0000) knlGS:0000000000000000
[    2.935412] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    2.935422] CR2: 00007fb174adfeb8 CR3: 00000001dec24005 CR4: 00000000003706e0
[    2.935433] Call Trace:
[    2.935444]  __clk_register+0x465/0x7e0
[    2.935459]  ? clk_hw_unregister+0x10/0x10
[    2.935471]  clk_hw_register+0x19/0x40
[    2.935481]  devm_clk_hw_register+0x41/0x80
[    2.935496]  rt5682_register_dai_clks.cold+0x63/0x16f [snd_soc_rt5682]
```
# test with GCC 9.3.0 and add more debug messages
```
[    2.868192] rt5682_i2c_probe, 1111 
[    2.868198] rt5682_register_dai_clks, i=0, init.name=rt5682-dai-wclk 
[    2.868212] rt5682 i2c-10EC5682:00: sysclk/dai not set correctly
[    2.868326] rt5682_register_dai_clks, dai_clk_hw->core=0x00000000c2fa6206
[    2.868333] rt5682_register_dai_clks, i=1, init.name=rt5682-dai-bclk 
[    2.868336] rt5682_register_dai_clks, RT5682_DAI_WCLK_IDX hw->core=0x00000000c2fa6206
[    2.868344] clk_core_get_parent_by_index, before clk_core_fill_parent_index, index=0
[    2.868347] clk_core_fill_parent_index, entry->hw=0x000000001cc04886 
[    2.868350] clk_core_fill_parent_index, entry->hw->core=0x0000000000000000 // got the wrong pointer address of clk_core of the parent of BCLK
[    2.868353] clk_core_get_parent_by_index, after clk_core_fill_parent_index
[    2.868359] clk_core_get_parent_by_index, before clk_core_fill_parent_index, index=0
[    2.868362] clk_core_fill_parent_index, entry->hw=0x000000001cc04886 
[    2.868365] clk_core_fill_parent_index, entry->hw->core=0x0000000000000000
[    2.868367] clk_core_get_parent_by_index, after clk_core_fill_parent_index
[    2.868502] rt5682_register_dai_clks, dai_clk_hw->core=0x000000004c754a03
[    2.868508] rt5682_i2c_probe, 222 
[    2.868692] cml_rt1011_rt5682 cml_rt1011_rt5682: sof_rt1011_quirk = f
```

**Comment by plbossart** [comment]
@shumingfan it's typical of race conditions, use a different compiler or add/remove traces modifies the timing. It's hard to find out what happens with such behavior....

File: sound/soc/codecs/rt5682.c [hunk 1]
```diff
@@ -2521,13 +2521,15 @@ static int rt5682_wclk_prepare(struct clk_hw *hw)
...
-	struct snd_soc_dapm_context *dapm =
-			snd_soc_component_get_dapm(component);
+	struct snd_soc_component *component;
+	struct snd_soc_dapm_context *dapm;
 
 	if (!rt5682_clk_check(rt5682))
 		return -EINVAL;
 
+	component = rt5682->component;
+	dapm = snd_soc_component_get_dapm(component);
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c:2531]: do you actually need this specific change @shumingfan ? I don't really see why it's necessary?

File: sound/soc/codecs/rt5682.c [hunk 2]
```diff
@@ -2557,13 +2559,15 @@ static void rt5682_wclk_unprepare(struct clk_hw *hw)
...
-	struct snd_soc_dapm_context *dapm =
-			snd_soc_component_get_dapm(component);
+	struct snd_soc_component *component;
+	struct snd_soc_dapm_context *dapm;
 
 	if (!rt5682_clk_check(rt5682))
 		return;
 
+	component = rt5682->component;
+	dapm = snd_soc_component_get_dapm(component);
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c:2569]: same here, is this change needed?

File: sound/soc/codecs/rt5682.c [hunk 3]
```diff
@@ -2754,20 +2757,22 @@ static int rt5682_bclk_set_rate(struct clk_hw *hw, unsigned long rate,
...
+	component = rt5682->component;
+
 	factor = rt5682_bclk_get_factor(rate, parent_rate);
 
 	for_each_component_dais(component, dai)
 		if (dai->id == RT5682_AIF1)
 			break;
 	if (!dai) {
-		dev_err(component->dev, "dai %d not found in component\n",
+		dev_err(rt5682->i2c_dev, "dai %d not found in component\n",
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c:2775]: same here?

File: sound/soc/codecs/rt5682.c [hunk 4]
```diff
@@ -2818,6 +2822,7 @@ static int rt5682_register_dai_clks(struct snd_soc_component *component)
 			init.parent_hws = &(const struct clk_hw *){
 				&rt5682->dai_clks_hw[RT5682_DAI_WCLK_IDX]
 			};
+			barrier();
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c:2825]: what is this trying to do?

**Review by plbossart — CHANGES_REQUESTED** [review]
@shumingfan Not fully clear on some of the changes where you replace component->dev but the result is the same.
Also you have a hidden barrier() that is not explained.

File: sound/soc/codecs/rt5682.c (hunk 1 above)
Comment (shumingfan) [reply sound/soc/codecs/rt5682.c:2531]: @plbossart Sorry, I forgot to explain more about this PR.
It will call the callback function when doing the devm_clk_hw_register function in the clk.c.
Like this 
https://github.com/thesofproject/linux/blob/e1e85cff35eb5c92ab618216c052bb2fb669f307/drivers/clk/clk.c#L3544
Due to moving DAI clock registration to i2c probe, the codec component doesn't ready to access.
To avoid the NULL pointer reference, the driver assigns the component variable after the rt5682_clk_check.

File: sound/soc/codecs/rt5682.c (hunk 4 above)
Comment (shumingfan) [reply sound/soc/codecs/rt5682.c:2825]: @plbossart 
I found the point where assigns the wrong pointer address in the clk_core_populate_parent_map function.
https://github.com/thesofproject/linux/blob/e1e85cff35eb5c92ab618216c052bb2fb669f307/drivers/clk/clk.c#L3769
And then, the 'general protection fault' issue happened when accessing the clk_core of clk_hw.
https://github.com/thesofproject/linux/blob/e1e85cff35eb5c92ab618216c052bb2fb669f307/drivers/clk/clk.c#L426

The 'parent_hws[i]' come from the codec driver, please check below.
https://github.com/thesofproject/linux/blob/e1e85cff35eb5c92ab618216c052bb2fb669f307/sound/soc/codecs/rt5682.c#L2818
If I try to print the address of 'init.parent_hws[i]' in the codec driver, the issue will not occur.
I feel the compiler optimization might cause this issue.
Therefore, I added the barrier() after the 'init.parent_hws' assignment and the test result is ok.

File: sound/soc/codecs/rt5682.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/codecs/rt5682.c:2531]: sorry @shumingfan, I really don't get the flow. 

rt5682->component is initialized in rt5682_probe(), which is a component probe. So if you use this field in a check done when doing the i2c probe, it is not initialized yet.

File: sound/soc/codecs/rt5682.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/codecs/rt5682.c:2825]: not getting the explanation, sorry @shumingfan

**Comment by XiaoyunWu6666** [comment]
hi @shumingfan , I've tested the PR #3130 with kmod-load-unload for 50 rounds on multiple CML platforms on Sep  29.
The test showed all PASS ... 
I manually run kmod load unload for 250 rounds on CML_HEL_RT5682 today, also got an all PASS

but not sure whether PR 3130 does fix #3108 or not , because this is an issue with rather low reproduction rate.

**Comment by plbossart** [comment]
I can't recall if this problem was ever root-caused/solved? @shumingfan @jack-cy-yu @ujfalusi can you refresh my memory?

**Comment by bardliao** [comment]
> I can't recall if this problem was ever root-caused/solved? @shumingfan @jack-cy-yu @ujfalusi can you refresh my memory?

@plbossart This PR is for https://github.com/thesofproject/linux/issues/3008. A version without `barrier();` is merged upstream and it seems solve https://github.com/thesofproject/linux/issues/3008. But we will see the `general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000` issue with that version. And as I know, we didn't root cause the `general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000` issue.

**Comment by ujfalusi** [comment]
> > I can't recall if this problem was ever root-caused/solved? @shumingfan @jack-cy-yu @ujfalusi can you refresh my memory?
> 
> @plbossart This PR is for #3008. A version without `barrier();` is merged upstream and it seems solve #3008. But we will see the `general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000` issue with that version. And as I know, we didn't root cause the `general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000` issue.

@bardliao, having a `barrier()` in a codec (i2c) driver is really unexpected in the first place, it is not even possible to have a general protection fault when talking to an external chip via i2c bus.

Do you have some longer logs or a way to reproduce it?

**Comment by bardliao** [comment]
> > > I can't recall if this problem was ever root-caused/solved? @shumingfan @jack-cy-yu @ujfalusi can you refresh my memory?
> > 
> > 
> > @plbossart This PR is for #3008. A version without `barrier();` is merged upstream and it seems solve #3008. But we will see the `general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000` issue with that version. And as I know, we didn't root cause the `general protection fault, probably for non-canonical address 0xd7a152bd29b38e00: 0000` issue.
> 
> @bardliao, having a `barrier()` in a codec (i2c) driver is really unexpected in the first place, it is not even possible to have a general protection fault when talking to an external chip via i2c bus.
> 
> Do you have some longer logs or a way to reproduce it?

No, I don't have any logs, but @shumingfan has. The issue can be reproduced when you apply 57589f82762e40bdaa975d840fa2bc5157b5be95 and build with gcc 9.3.0. And of cause you have to run the kernel on a device with rt5682 I2S codec.

**Comment by shumingfan** [comment]
Close this PR. Will revisit if necessary.
